# Lab Work 2: Text Processing: Preparation of texts

Use this notebook for the subsequence excecise's parts

## 6.2.1 Load the data and CountVectorize them
You will find a list of files in Ilias [sherlock.zip](https://www.ili.fh-aachen.de/goto_elearning_file_815003_download.html)
Download the zip file and adapt your next line accordingly.

In [1]:
import pandas as pd
import numpy as np

filenames = [r"C:/Users/alant/Sherlock.txt", 
             r"C:/Users/alant/Sherlock_blanched.txt",
             r"C:/Users/alant/Sherlock_black.txt",
             r"C:/Users/alant/Sherlock_blue.txt",
             r"C:/Users/alant/Sherlock_card.txt"]

Now we create a count Vectorizer. The parameter given tells the CountVectorizer that its methods shall operate on a list of filenames.

In [2]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer(input="filename")

After using the [CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html), I use the `fit_transform` function, which transforms the file names into a document-term matrix.

In [3]:
X = vectorizer.fit_transform(filenames).toarray()
X

array([[0, 1, 0, ..., 1, 0, 1],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [2, 0, 1, ..., 0, 1, 0],
       [0, 0, 0, ..., 0, 0, 0]])

### Create a DataFrame for simpler processing of data

In [4]:
df = pd.DataFrame(data=X, columns=vectorizer.get_feature_names_out())
df

,117,12,12s,13,131,13th,14th,15,16,1840,...,yourselves,youth,youthful,youths,yoxley,zealous,zenith,zest,zigzag,zoo
0,0,1,0,1,2,1,1,1,2,1,...,1,5,1,0,7,1,1,1,0,1
1,0,0,0,0,0,0,0,0,0,0,...,0,1,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,1,0,0,0,0,0,0,0,0
3,2,0,1,0,0,0,0,0,0,0,...,1,1,0,0,0,0,0,0,1,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,1,0,0,0,0,0,0


The output is a document-term matrix where the order of the words is ignored. Each row of the matrix represents a document (or a file) from your input. Each column represents a word in the entire vocabulary of all documents. In the cells, you will find the counter for the frequency of a word from the vocabulary in the document.

Now generate the Bag of Words with the CountVectorizer and check:
* the total number of different words
* the total number of words per document
* the total number of occurences of each word

### The total number of different words

I get the total number of different words by using `len(df.columns)` to return the number of columns in the DataFrame. I ignore upper and lower case and consider words such as `the` and `the` as not different.

In [5]:
diffnumWordsTotal = len(df.columns)
print("\nTotal number of different words: ", diffnumWordsTotal)


Total number of different words:  8879


There are a total of 8879 different words, whereby upper and lower case was ignored.  If there is a high number of dimensions, we should try to reduce the number of dimensions. In our case the total number of 8879 different words leads to high number of dimensions.

### The total number of words per document

I get the total number of words per document by calculating the sum of the words for each line of the DataFrame. A document is one row in our data set, so we take the sum of the rows 

In [6]:
numWordsPerDoc = df.sum(axis=1)
display("the total number of words per document", numWordsPerDoc)

'the total number of words per document'

0    107416
1      7258
2      7775
3      7497
4      8242
dtype: int64

The output shows the total number of words per document by clearly showing that document 1 `d0` has the most words.

### The total number of occurences of each word

I determine the total number of occurrences of each word by calculating the sum of the occurrences of each column. Each feature dimension is a single word, so I go over the columns.

In [7]:
numOccurencesPerWord = df.sum(axis=0)
display("the total number of occurences of each word", numOccurencesPerWord)

'the total number of occurences of each word'

117        2
12         1
12s        1
13         1
131        2
          ..
zealous    1
zenith     1
zest       1
zigzag     1
zoo        1
Length: 8879, dtype: int64

By reducing the dimension, it is most likely possible to make better statements regarding the occurrence of a word, as there are currently too many stop words or words with low information to make good statements.

## 6.2.2 Which word is occuring the most?

This must be done in three steps. Reason is, that the vectorizer.vocabulary_ is organized as a dictonary with the value indicating the position of the word in the array
1. Find out the highest count of a word
2. Find out the position of this count
3. Find out the word at this position

### Print the Vocabulary to represents the assignment of words to index positions in the vector

In [8]:
vocabulary_ = vectorizer.vocabulary_
display("vocabulary_= ",vocabulary_)

'vocabulary_= '

{'the': 7921,
 'return': 6563,
 'of': 5385,
 'sherlock': 7015,
 'holmes': 3843,
 'collection': 1528,
 'adventures': 190,
 'by': 1142,
 'sir': 7146,
 'arthur': 487,
 'conan': 1614,
 'doyle': 2454,
 'adventure': 189,
 'empty': 2687,
 'house': 3902,
 'it': 4319,
 'was': 8572,
 'in': 4038,
 'spring': 7411,
 'year': 8847,
 '1894': 18,
 'that': 7918,
 'all': 274,
 'london': 4711,
 'interested': 4239,
 'and': 342,
 'fashionable': 2997,
 'world': 8800,
 'dismayed': 2343,
 'murder': 5136,
 'honourable': 3858,
 'ronald': 6649,
 'adair': 150,
 'under': 8269,
 'most': 5101,
 'unusual': 8376,
 'inexplicable': 4115,
 'circumstances': 1411,
 'public': 6126,
 'has': 3701,
 'already': 293,
 'learned': 4565,
 'those': 7967,
 'particulars': 5580,
 'crime': 1898,
 'which': 8659,
 'came': 1171,
 'out': 5460,
 'police': 5852,
 'investigation': 4286,
 'but': 1132,
 'good': 3499,
 'deal': 2055,
 'suppressed': 7725,
 'upon': 8382,
 'occasion': 5365,
 'since': 7138,
 'case': 1241,
 'for': 3207,
 'prosecution': 

### Find out the highest count of a word

The variable `numOccurencesPerWord` is an array which contains the number of occurrences of each word. 

The [max()](https://pandas.pydata.org/docs/reference/api/pandas.Series.max.html) function finds the highest value in the array and stores this value in a new variable `highest_count_of_word`.

In [9]:
highest_count_of_word = numOccurencesPerWord.max(axis=0)
print("\nHighest count of a word:", highest_count_of_word, "\n")


Highest count of a word: 7975 



### Find out the word at this position

By using the [idmax()](https://pandas.pydata.org/docs/reference/api/pandas.Series.idxmax.html) function, I return the word with the highest occurrence. 

In this case the row label is the word I am searching for, and the maximum value is the highest count of a word

In [10]:
word_with_highest_occurence = numOccurencesPerWord.idxmax()
print("\nWord with the highest occurence:", word_with_highest_occurence, "\n")


Word with the highest occurence: the 



The word with the highest number of occurrences is the word `the` with a number of 7975 counts. I even expected this result, because `the` is one of the stop words that occur too often

### Find out the position of this count

After that, I use the variable in which the word with the highest number is stored as the key for the [vocabulary_](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) function and display the position of the word. 

In [11]:
position_of_word = vocabulary_[word_with_highest_occurence]
print("\nPosition of the word:", position_of_word, "\n")


Position of the word: 7921 



# 6.3 Improving using stop word, ngrams and tf-idf
The feature space is vast with nearly 9000 dimensions. Hence we should try to reduce the number of dimensions by:

1. use only words that have a mimimum occurence in all documents (minimal document frequency) min_df
2. remove stop words (like 'a', 'and', 'the') as they don't give valuable information for classification and/or 
3. remove words that occur in many documents (maximum document frequency) max_df 

Experiment with the values of min_df and max_df and see how the size of the vocabulary is changing.

Implement all three options and check for their separate outcome an their combinations

To reduce the number of dimensions, I remove all stop words so that the words that contain no information are removed. 

In addition, I tell the CountVectoriser with `min_df=3` and `max_df=0.75` that words that occur in less than 3 documents and words that occur in more than 75% of the documents are ignored.

To get some semantics, I create bigrams `ngram_range=(2,2)`, which are sequences of two consecutive words. 

In [12]:
# your code here
improvingVectorizer = CountVectorizer(input="filename", ngram_range=(2,2), stop_words="english",max_df=0.75, min_df=3)
Y = improvingVectorizer.fit_transform(filenames).toarray()
Y

array([[ 1, 12,  5,  6,  1,  1,  1,  5,  1,  1,  1,  1,  6,  1, 12,  2,
         1,  8,  1,  4,  4,  5,  2, 10,  3, 11,  1,  1,  1,  4,  3,  8,
         3,  3, 21,  2,  3,  2,  1,  1,  1,  8,  2,  1,  1,  3,  2,  1,
         1,  2,  1,  2,  4,  3,  1,  3,  2,  3,  1,  6,  4,  4, 10,  3,
         1,  2,  1,  2,  7, 13,  4,  3,  1, 11,  3,  1,  1,  1,  2,  0,
         1,  2,  1,  1,  2,  2,  1, 10,  3,  2, 10,  3,  1,  1,  1,  1,
         2,  2,  4,  2,  1,  1, 25,  1,  1,  1,  2,  8, 43],
       [ 1,  0,  2,  1,  0,  0,  1,  0,  1,  1,  1,  0,  1,  1,  0,  0,
         1,  0,  1,  0,  0,  3,  0,  2,  1,  0,  0,  0,  1,  0,  0,  1,
         0,  0,  0,  1,  0,  1,  0,  0,  0,  0,  2,  0,  1,  2,  0,  0,
         0,  0,  0,  1,  0,  1,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  1,  0,  0,  2,  0,  0,  0,  0,  0,  1,  1,  1,  0,  1,
         0,  0,  1,  2,  1,  0,  0,  0,  1,  0,  0,  0,  0,  0,  1,  0,
         1,  0,  2,  0,  0,  1,  0,  0,  1,  1,  1,  1,  2],
       [ 0,  5

In the following, I will use the same pattern to analyse the text as above before I adapted my CountVectorizer to reduce the number of dimensions ! 

In [13]:
df2 = pd.DataFrame(data=Y, columns=improvingVectorizer.get_feature_names_out())
df2

,able tell,asked holmes,bearded man,blue eyes,came examine,chair stared,closed book,closed door,come evening,come look,...,time ago,took heels,waited bit,watson said,wednesday tuesday,window mr,window way,wish know,yes said,young man
0,1,12,5,6,1,1,1,5,1,1,...,2,1,1,25,1,1,1,2,8,43
1,1,0,2,1,0,0,1,0,1,1,...,0,0,1,0,0,1,1,1,1,2
2,0,5,1,0,1,0,1,0,0,1,...,1,0,1,0,1,1,1,0,1,3
3,1,0,0,0,0,1,0,1,1,0,...,0,1,0,1,0,0,0,1,0,0
4,0,1,0,1,1,1,0,1,0,0,...,1,1,0,2,1,0,0,0,0,0


In [14]:
diffnumWordsTotal2 = len(df2.columns)
numwordsperdoc2 = df2.sum(axis=1)
numOccurencesPerWord2 = df2.sum(axis=0)
display("Total number of different words: ", diffnumWordsTotal2, "the total number of words per document", numwordsperdoc2, "the total number of occurences of each word", numOccurencesPerWord2)

'Total number of different words: '

109

'the total number of words per document'

0    417
1     53
2     98
3     52
4     49
dtype: int64

'the total number of occurences of each word'

able tell        3
asked holmes    18
bearded man      8
blue eyes        8
came examine     3
                ..
window mr        3
window way       3
wish know        4
yes said        10
young man       48
Length: 109, dtype: int64

As you can easily see, I have reduced the number of dimensions from, 9000 to 109. 

In fact, it makes a big difference if I leave stop words in or not. If I leave stop words in and do not change the other parameters, the dimension increases by 1500. 

However, if I only remove the parameter `max_df=0.75`, the number of dimensions increases by 21. In conclusion, it can be said that the stop words make up a large proportion of the words in the documents.

### Print the Vocabulary to represents the assignment of words to index positions in the vector

In [15]:
vocabulary_2 = improvingVectorizer.vocabulary_
display("vocabulary_= ",vocabulary_2)

'vocabulary_= '

{'good deal': 23,
 'sitting room': 90,
 'young man': 108,
 'watson said': 102,
 'holmes cried': 29,
 'moment later': 59,
 'making way': 52,
 'took heels': 100,
 'quite certain': 74,
 'mrs hudson': 62,
 'friend watson': 21,
 'half past': 25,
 'quarter hour': 73,
 'open door': 68,
 'blue eyes': 3,
 'holmes face': 31,
 'like man': 42,
 'holmes said': 34,
 'dressing gown': 17,
 'yes said': 107,
 'sake don': 81,
 'motive crime': 61,
 'ground floor': 24,
 'opened door': 69,
 'asked holmes': 1,
 'oh yes': 67,
 'shook head': 87,
 'far know': 19,
 'did know': 14,
 'little man': 45,
 'said little': 78,
 'know little': 37,
 'holmes examined': 30,
 'think mr': 98,
 'slip paper': 91,
 'tell said': 96,
 'wish know': 106,
 'single word': 88,
 'looked round': 51,
 'public house': 71,
 'dear sir': 12,
 'bearded man': 2,
 'man want': 55,
 'police station': 70,
 'morrow morning': 60,
 'sat deep': 82,
 'chair stared': 5,
 'come evening': 8,
 'holmes rising': 33,
 'hat head': 26,
 'mean say': 57,
 'knows m

In [16]:
highest_count_of_word2 = numOccurencesPerWord2.max(axis=0)
word_with_highest_occurence2 = numOccurencesPerWord2.idxmax()
position_of_word2 = vocabulary_2[word_with_highest_occurence2]
display("Highest count of a word: " + str(highest_count_of_word2), "Word with the highest occurence: " + str(word_with_highest_occurence2), "Position of the word: " + str(position_of_word2))

'Highest count of a word: 48'

'Word with the highest occurence: young man'

'Position of the word: 108'

After improving the bag of words, the highest count of a word was reduced from, 7975 to 48. 

By removing the stop words and using `min_df` and `max_df` with the `ngrams`, the word "the" no longer occurs, so that the word combination "young man" is the word with the highest occurrence.

This could indicate that it is a central term or a recurring term in the text.

# 6.4 Rescaling the data using term frequency inverse document frequency
Here, term frequency is the number of occurences of a term (word) $t$ in a document $d$. 

$\operatorname{tf}(t, d) = f_{t, d}$ 

Sometimes tf gets normalized to the length of $d$
The inverse document frequency idf is a measure on the amount of information a term t carries. Rare occurences of t leads to a high amount of information common occurence to a low amount of information. The idf is computed as 

$\text{idf}(t) = \log{\frac{1 + n}{1+\text{df}(t)}} + 1$

where $n$ is the total number of documents and $\text{df}(t)$ is the number of documents that contain the term $t$. Hence, the tf-idf is the product of the two terms:

$\text{tf-idf(t,d)}=\text{tf(t,d)} \cdot \text{idf(t)}$

scikit-learn supports this in the `TfidfTransformer`, when using the following parameters: `TfidfTransformer(norm='l2', use_idf=True, smooth_idf=True, sublinear_tf=False)`. Refer to the scikit documentation for the parameter sets and how this changes the formula.

Combining Bag of Words and tf-idf can be done using the `TfidfVectorizer`

In [17]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vect = TfidfVectorizer(input="filename", ngram_range=(2,2),stop_words="english", max_df=0.75, min_df=3)
X_tfidf = pd.DataFrame(tfidf_vect.fit_transform(filenames).toarray(), columns=tfidf_vect.get_feature_names_out())
X_tfidf

,able tell,asked holmes,bearded man,blue eyes,came examine,chair stared,closed book,closed door,come evening,come look,...,time ago,took heels,waited bit,watson said,wednesday tuesday,window mr,window way,wish know,yes said,young man
0,0.014316,0.171797,0.071582,0.085899,0.014316,0.014316,0.014316,0.071582,0.014316,0.014316,...,0.028633,0.014316,0.014316,0.357911,0.014316,0.014316,0.014316,0.028633,0.114531,0.615606
1,0.115470,0.000000,0.230940,0.115470,0.000000,0.000000,0.115470,0.000000,0.115470,0.115470,...,0.000000,0.000000,0.115470,0.000000,0.000000,0.115470,0.115470,0.115470,0.115470,0.230940
2,0.000000,0.428746,0.085749,0.000000,0.085749,0.000000,0.085749,0.000000,0.000000,0.085749,...,0.085749,0.000000,0.085749,0.000000,0.085749,0.085749,0.085749,0.000000,0.085749,0.257248
3,0.129099,0.000000,0.000000,0.000000,0.000000,0.129099,0.000000,0.129099,0.129099,0.000000,...,0.000000,0.129099,0.000000,0.129099,0.000000,0.000000,0.000000,0.129099,0.000000,0.000000
4,0.000000,0.108465,0.000000,0.108465,0.108465,0.108465,0.000000,0.108465,0.000000,0.000000,...,0.108465,0.108465,0.000000,0.216930,0.108465,0.000000,0.000000,0.000000,0.000000,0.000000


Each word now possesses a TF-IDF value when it appears in a document, or zero if it is absent in a document (row of the matrix).

# 6.4.1 Find maximum value for each of the features over dataset

In this section, we will find the maximum value for each feature in the dataset.

We generate a single value for each term, indicating its significance in the classification of a document

To make it a little clearer and more obvious, I have displayed the first 20 and the last 20 features with their maximum values to make the difference in value clearer.

In [18]:
# your code here
max_values = X_tfidf.max(axis=0)
sorted_by_tfidf = max_values.argsort()
feature_names = X_tfidf.columns
print("Features with lowest tfidf:\n{}".format(
    feature_names[sorted_by_tfidf[:20]]))
print("\ntf-idf values:\n{}".format(max_values.iloc[sorted_by_tfidf[:20]]))
print("\nFeatures with highest tfidf: \n{}".format(
    feature_names[sorted_by_tfidf[-20:]]))
print("\ntf-idf values:\n{}".format(max_values.iloc[sorted_by_tfidf[-20:]]))

Features with lowest tfidf:
Index(['came examine', 'door saw', 'holmes examined', 'far know',
       'head turned', 'holmes cried', 'motive crime', 'holmes rising',
       'murdered man', 'look face', 'll send', 'mean say', 'sailor man',
       'struck light', 'south american', 'pulled knife', 'time ago',
       'wednesday tuesday', 'sir good', 'little house'],
      dtype='object')

tf-idf values:
came examine         0.108465
door saw             0.108465
holmes examined      0.108465
far know             0.108465
head turned          0.108465
holmes cried         0.108465
motive crime         0.108465
holmes rising        0.108465
murdered man         0.108465
look face            0.108465
ll send              0.108465
mean say             0.108465
sailor man           0.108465
struck light         0.108465
south american       0.108465
pulled knife         0.108465
time ago             0.108465
wednesday tuesday    0.108465
sir good             0.108465
little house         0.11547

##### Features with lowest tf-idf:

The words listed have low TF-IDF values, which indicates that they occur frequently in the documents analysed and have a low information across all documents. They could be everyday words or stop words that are less important for understanding the text. In our case, stop words were removed so that represent typical phrases in the entire text corpus or are not very relevant with regard to classify a document.

##### Features mit höchstem tf-idf:

Words with high TF-IDF values are less common throughout the corpus, and are words that indicate high information.

##### General conclusions:

Words like "holmes examined" and "mean say" exhibit low TF-IDF values, suggesting that they are common phrases across the entire text corpus. The phrase "night went" stands out with a slightly higher TF-IDF value among the lower values, implying that this word pairing occurs less frequently and may offer more nuanced context. Terms, such as "mrs hudson" and "watson," show higher TF-IDF values, signifying their heightened presence in specific sections of the texts.